### PR TITLE
🐛 Use syncmap for the resource cache

### PR DESCRIPTION
**Pull request by zhujian7**
## Related issue(s)

https://github.com/open-cluster-management-io/ocm/issues/1020

Fixes #

### CODE REVIEW

File: pkg/work/spoke/apply/resource_cache_test.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	}
+	cache.UpdateCachedResourceMetadata(obj, obj3)
+	if !cache.SafeToSkipApply(obj, obj3) {
+		t.Fatal("expected SafeToSkipApply to return true after updating cache with new resource version")
+	}
+}
+
+func TestCurrentReadWriteCache(t *testing.T) {
+
+	cache := resourceapply.NewResourceCache()
```
Comment (zhujian7) [inline pkg/work/spoke/apply/resource_cache_test.go]: the openshift library-go resourceCache failed with error:
```
goroutine 9 [running]:
internal/runtime/maps.fatal({0x1061e5bc5?, 0xf?})
	/opt/homebrew/Cellar/go/1.24.3/libexec/src/runtime/panic.go:1058 +0x20
github.com/openshift/library-go/pkg/operator/resource/resourceapply.(*resourceCache).SafeToSkipApply(0x14000231ce8, {0x106bcfeb8?, 0x14000231cf0?}, {0x106bcfeb8, 0x14000231cf0})
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/resource_cache.go:148 +0xb4
open-cluster-management.io/ocm/pkg/work/spoke/apply.TestCurrentReadWriteCache.func1()
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/pkg/work/spoke/apply/resource_cache_test.go:84 +0xb0
created by open-cluster-management.io/ocm/pkg/work/spoke/apply.TestCurrentReadWriteCache in goroutine 8
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/pkg/work/spoke/apply/resource_cache_test.go:79 +0x138
```
code:
```
	if cached, exists := c.cache[cacheKey]; exists {    // this line
		versionMatch = cached.resourceVersion == resourceVersion
		hashMatch = cached.resourceHash == resourceHash
		if versionMatch && hashMatch {
			klog.V(4).Infof("found matching resourceVersion & manifest hash")
			return true
		}
	}
```

Comment (zhujian7) [reply pkg/work/spoke/apply/resource_cache_test.go]: even with the syncmap, the test still failed with the inner unstructured NestedFieldNocopy func:
```
goroutine 104 [running]:
internal/runtime/maps.fatal({0x103a9dbe5?, 0x1026bd088?})
	/opt/homebrew/Cellar/go/1.24.3/libexec/src/runtime/panic.go:1058 +0x20
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.NestedFieldNoCopy(0x140006dbbb8?, {0x140006dbc58, 0x2, 0x2})
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:62 +0xac
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.NestedString(0x14000192ccd?, {0x140006dbc58, 0x2, 0x2})
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:76 +0x28
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.getNestedString(0x14000192ccd?, {0x140006dbc58?, 0x103a76cbb?, 0xa?})
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:302 +0x20
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GetName(0x14000922628?)
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:251 +0x5c
open-cluster-management.io/ocm/pkg/work/spoke/apply.getResourceMetadata({0x104487eb8, 0x14000922628})
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/pkg/work/spoke/apply/resource_cache.go:65 +0x3d8
open-cluster-management.io/ocm/pkg/work/spoke/apply.(*resourceCache).SafeToSkipApply(0x14000922620, {0x104487eb8?, 0x14000922628?}, {0x104487eb8, 0x14000922628})
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/pkg/work/spoke/apply/resource_cache.go:125 +0x5c
open-cluster-management.io/ocm/pkg/work/spoke/apply.TestCurrentReadWriteCache.func1()
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/pkg/work/spoke/apply/resource_cache_test.go:83 +0xb0
created by open-cluster-management.io/ocm/pkg/work/spoke/apply.TestCurrentReadWriteCache in goroutine 10
	/Users/jiazhu/go/src/open-cluster-management.io/ocm/pkg/work/spoke/apply/resource_cache_test.go:78 +0x140
```

```
func NestedFieldNoCopy(obj map[string]interface{}, fields ...string) (interface{}, bool, error) {
	var val interface{} = obj

	for i, field := range fields {
		if val == nil {
			return nil, false, nil
		}
		if m, ok := val.(map[string]interface{}); ok {
			val, ok = m[field]         //  this line
			if !ok {
				return nil, false, nil
			}
		} else {
			return nil, false, fmt.Errorf("%v accessor error: %v is of the type %T, expected map[string]interface{}", jsonPath(fields[:i+1]), val, val)
		}
	}
	return val, true, nil
}
```

File: vendor/golang.org/x/sync/syncmap/map.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import "sync" // home to the standard library's sync.map implementation as of Go 1.9
+
+// Map is a concurrent map with amortized-constant-time loads, stores, and deletes.
+// It is safe for multiple goroutines to call a Map's methods concurrently.
+//
+// The zero Map is valid and empty.
+//
+// A Map must not be copied after first use.
+type Map = sync.Map
```
Comment (qiujian16) [inline vendor/golang.org/x/sync/syncmap/map.go]: seems we can directly use sync.Map?

**Review by qiujian16 — COMMENTED** [review]
### pkg/work/spoke/apply/resource_cache.go

```go
@@ +10 @@
+type cachedVersionKey struct {
+	name      string
+	namespace string
+	kind      schema.GroupKind
+}

// Consider making these fields unexported unless there's a specific reason for them to be public.

@@ +16 @@
+type cachedResource struct {
+	resourceHash, resourceVersion string
+}

// Same here, consider making these fields unexported unless necessary.

@@ +25 @@
+type resourceCache struct {
+	cache sync.Map // use syncmap for concurrent access
+}

// Consider adding a comment explaining why sync.Map is chosen over a regular map with mutex.

@@ +32 @@
+func NewResourceCache() *resourceCache {
+	return &resourceCache{
+		cache: sync.Map{},
+	}
+}

// This looks good. Ensure consistency with naming convention in the package (camelCase or PascalCase).

@@ +61 @@
+		metadata, err := meta.Accessor(obj)
+		if err != nil {
+			return schema.GroupKind{}, "", "", "", err
+		}
+		if metadata == nil || reflect.ValueOf(metadata).IsNil() {
+			return schema.GroupKind{}, "", "", "", fmt.Errorf("object has no metadata")
+		}
+		resourceHash := hashOfResourceStruct(obj)

// Ensure `hashOfResourceStruct` is defined somewhere in the codebase, otherwise it will lead to a compile-time error.

@@ +68 @@
+		currKind := getCoreGroupKind(obj)
+		if currKind != nil {
+			kind = *currKind
+		}
+	}
+	if len(kind.Kind) == 0 {
+		return schema.GroupKind{}, "", "", "", fmt.Errorf("unable to determine GroupKind of %T", obj)
+	}

// This can be simplified by directly returning the error if `getCoreGroupKind(obj)` returns nil.

@@ +90 @@
+func getResourceVersion(obj runtime.Object) (string, error) {
+	if obj == nil {
+		return "", fmt.Errorf("nil object has no resourceVersion")
+	}
+	metadata, err := meta.Accessor(obj)
+	if err != nil {
+		return "", err
+	}
+	if metadata == nil || reflect.ValueOf(metadata).IsNil() {
+		return "", fmt.Errorf("object has no metadata")
+	}
+	rv := metadata.GetResourceVersion()
+	if len(rv) == 0 {
+		return "", fmt.Errorf("missing resourceVersion")
+	}

// The checks for metadata being nil twice are redundant. You can combine them into a single check.

@@ +120 @@
+	// retrieve kind, sometimes this can be done via the accesor, sometimes not (depends on the type)
+	kind := schema.GroupKind{}
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	if len(gvk.Kind) > 0 {
+		kind = gvk.GroupKind()
+	} else {
+		if currKind := getCoreGroupKind(obj); currKind != nil {
+			kind = *currKind
+		}
+	}

// Typo in "accessor". Also, consider simplifying the nested if statement.

@@ +182 @@
+func (c *resourceCache) SafeToSkipApply(required runtime.Object, existing runtime.Object) bool {
+	if c == nil {
+		return false
+	}
+	if required == nil || existing == nil {
+		return false
+	}
+	kind, name, namespace, resourceHash, err := getResourceMetadata(required)
+	if err != nil {
+		return false
+	}
+	cacheKey := cachedVersionKey{
+		name:      name,
+		namespace: namespace,
+		kind:      kind,
+	}

// The logic here is clear but consider logging or metrics for errors, especially if they occur frequently.

@@ +190 @@
+	resourceVersion, err := getResourceVersion(existing)
+	if err != nil {
+		return false
+	}

// Similar comment as above.

@@ +231 @@
+func getCoreGroupKind(obj runtime.Object) *schema.GroupKind {
+	switch obj.(type) {
+	case *corev1.Namespace:
+		return &schema.GroupKind{
+			Kind: "Namespace",
+		}
+	case *corev1.Service:
+		return &schema.GroupKind{
+			Kind: "Service",
+		}
+	case *corev1.Pod:
+		return &schema.GroupKind{
+			Kind: "Pod",
+		}
+	case *corev1.ServiceAccount:
+		return &schema.GroupKind{
+			Kind: "ServiceAccount",
+		}
+	case *corev1.ConfigMap:
+		return &schema.GroupKind{
+			Kind: "ConfigMap",
+		}
+	case *corev1.Secret:
+		return &schema.GroupKind{
+			Kind: "Secret",
+		}
+	default:
+		return nil
+	}
+}
```

### pkg/work/spoke/apply/resource_cache_test.go

```go
@@ +11 ##
+	if cache == nil {
+		t.Fatal("expected non-nil resource cache")
+	}

// Test naming consistency is vital for readability.

@@ +14 ##
+	obj := &unstructured.Unstructured{}
+	obj.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "apps",
+		Version: "v1",
+		Kind:    "Deployment",
+	})
+	obj.SetResourceVersion("12345")

// Consider breaking up large tests into smaller ones for better isolation.

@@ +21 ##
+	obj.SetName("test")
+	obj.SetNamespace("default")
+	// Test UpdateCachedResourceMetadata
+	cache.UpdateCachedResourceMetadata(obj, obj)

// Comments should describe the expected outcome, not what's being executed.

@@ +24 ##
+	// Test SafeToSkipApply
+	if !cache.SafeToSkipApply(obj, obj) {
+		t.Fatal("expected SafeToSkipApply to return true for identical objects")
+	}

// Ensure all tests cover edge cases.

@@ +33 ##
+	if cache.SafeToSkipApply(obj, obj2) {
+		t.Fatal("expected SafeToSkipApply to return false for different objects")
+	}

// Consider adding more descriptive names for variables to improve test readability.

@@ +45 ##
+	if !cache.SafeToSkipApply(obj, obj3) {
+		t.Fatal("expected SafeToSkipApply to return true after updating cache with new resource version")
+	}
+}
```

### pkg/work/spoke/apply/update_apply.go

```go
@@ -36 +36 @@
-		staticResourceCache: resourceapply.NewResourceCache(),
+		staticResourceCache: NewResourceCache(),
```

- Ensure `NewResourceCache` is exported and accessible here.

### pkg/work/spoke/apply/update_apply_test.go

```go
@@ -4 +3 @@
-	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	// github.com/openshift/library-go/pkg/operator/resource/resourceapply is removed.

@@ -325 +324 @@
-			cache := resourceapply.NewResourceCache()
+			cache := NewResourceCache()
```

- Ensure `NewResourceCache` is exported and accessible here.

### General Considerations

- Follow consistent naming conventions throughout the codebase.
- Improve logging to include more context when debugging.
- Ensure all functions are properly documented with godoc comments.
- Break down large functions and tests into smaller, more manageable pieces where possible.
- Validate assumptions in your code through additional unit tests.
